File: sdk/src/spec.ts
// Spec types for Relayflows — data, not code (RFC-0001 settled decision #5).
//
// Mirrors RFC-0001 §1's ladder. Every rung is a legal relayflow:
//   deterministic step          # a pure script — no LLM anywhere (legal)
//     + llm step                # a bare model call — prompt in, verified output out
//       + agent step            # a harnessed agent in a workspace — artifact + diff + trajectory
//
// Step types are exactly `deterministic | llm | agent` (AGENTS.md rule 7).
// Zero-agent flows are legal: a spec with only `deterministic` (and/or `llm`)
// steps is valid. Nothing here requires an `agent` step.

import type { JsonOutputSchema } from './output-schema.js';
export type { JsonOutputSchema } from './output-schema.js';

/** The three rungs of the ladder (RFC §1; AGENTS.md rule 7). */
export type StepType = 'deterministic' | 'llm' | 'agent';

/**
 * Verification is control flow, not decoration (kernel DESIGN.md §3).
 * v0 gates are deterministic so verification is kernel-side and replayable.
 */
export type VerificationGateType = 'exit_code' | 'output_contains' | 'json_schema';

/**
 * `exit_code == 0` — the implicit gate for deterministic steps. v0 judges
 * exactly zero (kernel DESIGN.md §4); it is not configurable, so this gate
 * carries no parameters. Writing it explicitly is allowed and compiles to the
 * same kernel spec as omitting it.
 */
export interface ExitCodeGate {
  type: 'exit_code';
}

/** Step output (stdout_tail / llm value, stringified) contains `value`. */
export interface OutputContainsGate {
  type: 'output_contains';
  value: string;
}

/** Step output validates against a JSON Schema. Used for `llm` structured output. */
export interface JsonSchemaGate {
  type: 'json_schema';
  schema: boolean | Record<string, unknown>;
}

export type VerificationSpec = ExitCodeGate | OutputContainsGate | JsonSchemaGate;
export type OutputVerificationSpec = OutputContainsGate | JsonSchemaGate;

/**
 * Agent-step recovery modes (RFC Appendix A rule 4). Default is `reset`.
 * `deterministic` / `llm` steps have no workspace, so these do not apply.
 */
export type RecoveryMode = 'reset' | 'inspect' | 'manual';

/**
 * Declared mutable surfaces for an agent step (RFC Appendix A rule 1).
 * Anything undeclared is outside the contract and outside the step's
 * permissions (gate 8 makes this enforceable, not advisory).
 */
export interface WorkspaceSurface {
  /** Relayfile mount path, or a named worktree. */
  surface: string;
}

export interface StreamSurface {
  /** Durable channel the agent may write (kernel DESIGN.md §1.8). */
  stream: string;
}

export interface AgentSurfaces {
  workspace?: WorkspaceSurface[];
  streams?: StreamSurface[];
  /** Integration writeback paths — mount writes per gate 6. */
  external?: string[];
}

/** Permission model for an agent step (gate 8). `readonly` provably cannot write. */
export interface PermissionsSpec {
  fileGlobs?: string[];
  networkAllowlist?: string[];
  accessPreset?: 'readonly' | 'readwrite';
}

/**
 * Budget envelope. Per decision #10 every token has exactly one owner: an
 * injected context pack spends the consuming step's budget. Money is a decimal
 * string at the boundary (kernel DESIGN.md §1: no floats for money); tokens
 * are integers.
 */
export interface BudgetSpec {
  maxTokensIn?: number;
  maxTokensOut?: number;
  /** Decimal string, e.g. "1.50". */
  maxDollars?: string;
}

/** Fields shared by every step on the ladder. */
export interface BaseStepSpec {
  /** Stable step identity; journaled as `step_id` and hashed into the idempotency key. */
  id: string;
  type: StepType;
  /** Step dependencies — a step runs only after these complete. */
  dependsOn?: string[];
  /** Semantic retry bound (kernel DESIGN.md §1.2 `max_iterations`). Default 1. */
  maxIterations?: number;
}

/**
 * Rung 1 — a pure script. Executed by the `relayflowd` binary: spawn command,
 * capture stdout/exit code. Output = `{exit_code, stdout_tail}`. Gate-1
 * deterministic steps are pure (no pins).
 */
export interface DeterministicStepSpec extends BaseStepSpec {
  type: 'deterministic';
  command: string;
  /** Omit to get the implicit `exit_code` gate. */
  verification?: VerificationSpec;
}

/**
 * Rung 2 — a bare model call. No workspace, output is a value. The kernel
 * never calls a model: it dispatches to an attached SDK worker (§5) which
 * returns `{output, usage}`; the kernel then runs the verification gate.
 */
export interface LlmStepSpec extends BaseStepSpec {
  type: 'llm';
  prompt: string;
  verification?: OutputVerificationSpec;
  model?: string;
  /** Inert preflight declaration; overrides the flow/project CLI default. */
  cli?: string;
  /**
   * Structured-output authoring sugar. Compiles to the existing `json_schema`
   * verification primitive and is removed before the kernel boundary.
   */
  output?: JsonOutputSchema;
}

/**
 * Rung 3 — a harnessed agent in a workspace. Dispatched like `llm`, plus
 * Appendix A in full: pins declared workspace revisions and stream offsets;
 * every writeback is a journaled `effect.recorded` deduped by
 * `(step_id, idempotency_key, surface_path)`.
 */
export interface AgentStepSpec extends BaseStepSpec {
  type: 'agent';
  instruction: string;
  verification?: OutputVerificationSpec;
  /** Inert preflight declaration; overrides the flow/project CLI default. */
  cli?: string;
  /**
   * Model the declared CLI must use. Raw Claude/Codex adapters receive their
   * real model flag; an identified Relayflows wrapper receives
   * `RELAYFLOW_MODEL`. Declared here so the choice is journaled with the step
   * instead of being ambient host state.
   */
  model?: string;
  surfaces?: AgentSurfaces;
  recoveryMode?: RecoveryMode;
  permissions?: PermissionsSpec;
  /**
   * Structured-output authoring sugar. A successful CLI JSON object is the parsed
   * value; the kernel persists it only after `json_schema` verification.
   */
  output?: JsonOutputSchema;
}

export type StepSpec = DeterministicStepSpec | LlmStepSpec | AgentStepSpec;

/**
 * Reusable authoring declaration for an agent CLI/model pair. Both fields are
 * required so selecting a named agent can never inherit a host model. The
 * compiler lowers these values into the selected kernel agent step at the
 * journal boundary; the kernel never receives this map or a new step field.
 */
export interface NamedAgentSpec {
  cli: string;
  model: string;
}

/** Inert gate-1 trigger declaration. Matching and dispatch belong to gate 2. */
export interface TriggerSpec {
  id: string;
  /** Executor registration required before this trigger may start a run. */
  executor: string;
}

/**
 * A Relayflow spec in the authoring shape — the composable unit (RFC settled
 * decision #5). Schema-validated, diffable, signable (gate 8), and emittable
 * by a step (gate 9 self-authoring). What the kernel inlines in `run.spawned`
 * is this spec mapped to the kernel dialect (`toKernelSpec`).
 */
export interface FlowSpec {
  /** Spec schema semver (RFC §7). Compilers always emit latest. */
  version: string;
  name?: string;
  description?: string;
  /** Inert preflight default for llm/agent steps that do not declare a CLI. */
  cli?: string;
  /** Named authoring declarations. Compiled into agent steps, never journaled as a new primitive. */
  agents?: Record<string, NamedAgentSpec>;
  /** Declarations checked by preflight; gate 1 never dispatches them. */
  triggers?: TriggerSpec[];
  steps: StepSpec[];
  budget?: BudgetSpec;
}

/** Current spec schema version emitted by this SDK. */
export const SPEC_SCHEMA_VERSION = '0.1.0';

// --- The kernel dialect ------------------------------------------------------
//
// The authoring types above are TypeScript-idiomatic (camelCase, tagged
// verification sugar). The *boundary artifact* is singular: the kernel's spec
// dialect — snake_case keys, semver `version`, flat v0 verification, defaults
// materialized. `toKernelSpec` (compile.ts) maps authoring → kernel; parity
// with `kernel/relayflowd-core/src/spec.rs` is pinned bit-for-bit by
// `tests/spec-parity.test.ts` and the kernel's `tests/spec_parity.rs` over the
// shared `testdata/` fixture.

export interface KernelRetryPolicy {
  initial_backoff_ms: number;
  max_backoff_ms: number;
  multiplier: number;
  jitter_percent: number;
}

/**
 * Flat v0 gates (kernel DESIGN.md §4): `exit_code == 0` is implicit for
 * deterministic steps; these two are optional and combinable. An empty object
 * means "implicit gates only".
 */
export interface KernelVerificationSpec {
  output_contains?: string;
  json_schema?: boolean | Record<string, unknown>;
}

export interface KernelStepCommon {
  id: string;
  depends_on: string[];
  max_iterations: number;
  retry: KernelRetryPolicy;
  verification: KernelVerificationSpec;
}

export interface KernelDeterministicStep extends KernelStepCommon {
  type: 'deterministic';
  command: string;
  timeout_ms?: number;
}

export interface KernelLlmStep extends KernelStepCommon {
  type: 'llm';
  prompt: string;
  model?: string;
  cli?: string;
}

export interface KernelAgentSurfaces {
  workspace?: { surface: string }[];
  streams?: { stream: string }[];
  external?: string[];
}

export interface KernelPermissionsSpec {
  file_globs?: string[];
  network_allowlist?: string[];
  access_preset?: 'readonly' | 'readwrite';
}

export interface KernelAgentStep extends KernelStepCommon {
  type: 'agent';
  instruction: string;
  cli?: string;
  model?: string;
  recovery_mode: RecoveryMode;
  surfaces?: KernelAgentSurfaces;
  permissions?: KernelPermissionsSpec;
}

export type KernelStepSpec = KernelDeterministicStep | KernelLlmStep | KernelAgentStep;

export interface KernelBudgetSpec {
  max_tokens_in?: number;
  max_tokens_out?: number;
  max_dollars?: string;
}

export interface KernelTriggerSpec {
  id: string;
  executor: string;
}

/** The compiled spec as the kernel parses, journals, and hashes it. */
export interface KernelRunSpec {
  version: string;
  name?: string;
  description?: string;
  cli?: string;
  triggers?: KernelTriggerSpec[];
  steps: KernelStepSpec[];
  budget?: KernelBudgetSpec;
}

File: kernel/relayflowd-core/src/spec.rs
//! Run specs — the composable unit (RFC-0001 settled decision #5), one dialect.
//!
//! This is the single spec shape at the SDK↔kernel boundary: snake_case keys,
//! semver `version`, flat v0 verification (`output_contains` / `json_schema`;
//! `exit_code == 0` is implicit for deterministic steps — kernel DESIGN.md §4).
//! The SDK compiler emits exactly this shape; parity is pinned bit-for-bit by
//! `tests/spec_parity.rs` against `testdata/hello-ladder.spec.canonical.json`.
//!
//! Parsing is fail-closed (AGENTS.md rule 4): [`RunSpec::parse`] rejects any
//! unknown field, so a misspelled `verification` key can never silently drop a
//! gate. Nested structs carry `deny_unknown_fields`; step objects (which use
//! `#[serde(flatten)]`, where serde cannot enforce it) are checked explicitly.

use std::collections::{BTreeMap, BTreeSet};

use serde::{Deserialize, Serialize};
use serde_json::Value;
use thiserror::Error;

/// The spec schema version this kernel reads and writes (semver, RFC §7).
pub const SPEC_VERSION: &str = "0.1.0";

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct RunSpec {
    #[serde(default = "default_spec_version")]
    pub version: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Default CLI declaration for llm/agent steps. `flows check` resolves
    /// step → flow → project config. The kernel treats this as inert data;
    /// `run.start` does not invoke the surface preflight in gate 1.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cli: Option<String>,
    /// Inert gate-1 declarations. Matching and dispatch belong to gate 2.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub triggers: Vec<TriggerSpec>,
    #[serde(default)]
    pub steps: Vec<StepSpec>,
    /// Budget envelope (RFC settled decision #10). Carried and journaled from
    /// gate 1; enforced when llm/agent dispatch lands — the deterministic rung
    /// spends zero tokens.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub budget: Option<BudgetSpec>,
}

fn default_spec_version() -> String {
    SPEC_VERSION.to_owned()
}

impl RunSpec {
    /// Fail-closed parse: reject unknown fields everywhere before
    /// deserializing. `#[serde(flatten)]` on [`StepSpec`] prevents serde-level
    /// `deny_unknown_fields` for step objects, so their key sets are checked
    /// here; every non-flattened struct denies unknown fields via serde.
    pub fn parse(value: &Value) -> Result<Self, SpecError> {
        reject_unknown_step_fields(value)?;
        serde_json::from_value(value.clone())
            .map_err(|error| SpecError::Malformed(error.to_string()))
    }

    pub fn validate(&self) -> Result<(), SpecError> {
        if self.version != SPEC_VERSION {
            return Err(SpecError::UnsupportedVersion(self.version.clone()));
        }

        if self.cli.as_ref().is_some_and(|cli| cli.trim().is_empty()) {
            return Err(SpecError::EmptyCli);
        }

        let mut trigger_ids = BTreeSet::new();
        for trigger in &self.triggers {
            if trigger.id.trim().is_empty() || trigger.executor.trim().is_empty() {
                return Err(SpecError::InvalidTrigger(trigger.id.clone()));
            }
            if !trigger_ids.insert(trigger.id.clone()) {
                return Err(SpecError::DuplicateTrigger(trigger.id.clone()));
            }
        }

        let mut ids = BTreeSet::new();
        for step in &self.steps {
            if step.id.trim().is_empty() {
                return Err(SpecError::EmptyStepId);
            }
            if !ids.insert(step.id.clone()) {
                return Err(SpecError::DuplicateStep(step.id.clone()));
            }
            if step.max_iterations == 0 {
                return Err(SpecError::ZeroIterations(step.id.clone()));
            }
            let cli = match &step.kind {
                StepKind::Llm { cli, .. } | StepKind::Agent { cli, .. } => cli,
                StepKind::Deterministic { .. } => &None,
            };
            if cli.as_ref().is_some_and(|value| value.trim().is_empty()) {
                return Err(SpecError::EmptyStepCli(step.id.clone()));
            }
            step.retry.validate(&step.id)?;
        }

        for step in &self.steps {
            for dependency in &step.depends_on {
                if dependency == &step.id {
                    return Err(SpecError::DependencyCycle(step.id.clone()));
                }
                if !ids.contains(dependency) {
                    return Err(SpecError::UnknownDependency {
                        step: step.id.clone(),
                        dependency: dependency.clone(),
                    });
                }
            }
        }

        let dependencies = self
            .steps
            .iter()
            .map(|step| (step.id.as_str(), step.depends_on.as_slice()))
            .collect::<BTreeMap<_, _>>();
        let mut visiting = BTreeSet::new();
        let mut visited = BTreeSet::new();
        for id in &ids {
            visit(id, &dependencies, &mut visiting, &mut visited)?;
        }
        Ok(())
    }

    pub fn step(&self, id: &str) -> Option<&StepSpec> {
        self.steps.iter().find(|step| step.id == id)
    }
}

const STEP_COMMON_FIELDS: &[&str] = &[
    "id",
    "type",
    "depends_on",
    "max_iterations",
    "retry",
    "verification",
];
const STEP_DETERMINISTIC_FIELDS: &[&str] = &["command", "timeout_ms"];
const STEP_LLM_FIELDS: &[&str] = &["prompt", "model", "cli"];
const STEP_AGENT_FIELDS: &[&str] = &[
    "instruction",
    "cli",
    "recovery_mode",
    "surfaces",
    "permissions",
];

fn reject_unknown_step_fields(value: &Value) -> Result<(), SpecError> {
    let Some(steps) = value.get("steps").and_then(Value::as_array) else {
        return Ok(()); // shape errors surface from serde with their own message
    };
    for (index, step) in steps.iter().enumerate() {
        let Some(object) = step.as_object() else {
            continue;
        };
        let kind_fields = match object.get("type").and_then(Value::as_str) {
            Some("deterministic") => STEP_DETERMINISTIC_FIELDS,
            Some("llm") => STEP_LLM_FIELDS,
            Some("agent") => STEP_AGENT_FIELDS,
            // Missing/unknown type is rejected by serde's tagged-enum error.
            _ => continue,
        };
        for key in object.keys() {
            if !STEP_COMMON_FIELDS.contains(&key.as_str()) && !kind_fields.contains(&key.as_str()) {
                return Err(SpecError::UnknownField {
                    at: format!("steps[{index}]"),
                    field: key.clone(),
                });
            }
        }
    }
    Ok(())
}

fn visit<'a>(
    id: &'a str,
    dependencies: &BTreeMap<&'a str, &'a [String]>,
    visiting: &mut BTreeSet<&'a str>,
    visited: &mut BTreeSet<&'a str>,
) -> Result<(), SpecError> {
    if visited.contains(id) {
        return Ok(());
    }
    if !visiting.insert(id) {
        return Err(SpecError::DependencyCycle(id.to_owned()));
    }
    for dependency in dependencies.get(id).copied().unwrap_or_default() {
        visit(dependency, dependencies, visiting, visited)?;
    }
    visiting.remove(id);
    visited.insert(id);
    Ok(())
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct StepSpec {
    pub id: String,
    #[serde(default)]
    pub depends_on: Vec<String>,
    #[serde(default = "default_max_iterations")]
    pub max_iterations: u32,
    #[serde(default)]
    pub retry: RetryPolicy,
    #[serde(default)]
    pub verification: VerificationSpec,
    #[serde(flatten)]
    pub kind: StepKind,
}

fn default_max_iterations() -> u32 {
    1
}

impl StepSpec {
    pub fn step_type(&self) -> StepType {
        match self.kind {
            StepKind::Deterministic { .. } => StepType::Deterministic,
            StepKind::Llm { .. } => StepType::Llm,
            StepKind::Agent { .. } => StepType::Agent,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum StepKind {
    Deterministic {
        command: CommandSpec,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        timeout_ms: Option<u64>,
    },
    Llm {
        prompt: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        model: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        cli: Option<String>,
    },
    Agent {
        instruction: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        cli: Option<String>,
        #[serde(default)]
        recovery_mode: RecoveryMode,
        /// Declared mutable surfaces (RFC Appendix A rule 1) — names only.
        /// Revision/offset *pins* are runtime facts journaled per attempt
        /// (Appendix A rule 2), never spec fields.
        #[serde(default, skip_serializing_if = "AgentSurfaces::is_empty")]
        surfaces: AgentSurfaces,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        permissions: Option<PermissionsSpec>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum CommandSpec {
    Shell(String),
    Argv(Vec<String>),
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum StepType {
    Deterministic,
    Llm,
    Agent,
}

#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum RecoveryMode {
    #[default]
    Reset,
    Inspect,
    Manual,
}

/// Declared state surfaces for an agent step (RFC Appendix A rule 1):
/// workspace mounts/worktrees, writable streams, and external writeback paths.
#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct AgentSurfaces {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub workspace: Vec<WorkspaceSurface>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub streams: Vec<StreamSurface>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub external: Vec<String>,
}

impl AgentSurfaces {
    pub fn is_empty(&self) -> bool {
        self.workspace.is_empty() && self.streams.is_empty() && self.external.is_empty()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct WorkspaceSurface {
    pub surface: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct StreamSurface {
    pub stream: String,
}

/// Permission model for an agent step (gate 8). Carried as data in gate 1;
/// enforcement lands with agent dispatch.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct PermissionsSpec {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub file_globs: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub network_allowlist: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub access_preset: Option<AccessPreset>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AccessPreset {
    Readonly,
    Readwrite,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct TriggerSpec {
    pub id: String,
    pub executor: String,
}

/// Budget envelope: tokens are integers; money is a decimal string, never a
/// float (kernel DESIGN.md §1).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct BudgetSpec {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_tokens_in: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_tokens_out: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_dollars: Option<String>,
}

/// v0 verification gates (kernel DESIGN.md §4): `exit_code == 0` is implicit
/// for deterministic steps; these two are optional and combinable. Unknown
/// keys are a parse error — verification is control flow, and a dropped gate
/// is a fail-open bug, not a default.
#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
#[serde(deny_unknown_fields)]
pub struct VerificationSpec {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub output_contains: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub json_schema: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct RetryPolicy {
    #[serde(default = "default_initial_backoff_ms")]
    pub initial_backoff_ms: u64,
    #[serde(default = "default_max_backoff_ms")]
    pub max_backoff_ms: u64,
    #[serde(default = "default_multiplier")]
    pub multiplier: u32,
    #[serde(default = "default_jitter_percent")]
    pub jitter_percent: u8,
}

impl Default for RetryPolicy {
    fn default() -> Self {
        Self {
            initial_backoff_ms: default_initial_backoff_ms(),
            max_backoff_ms: default_max_backoff_ms(),
            multiplier: default_multiplier(),
            jitter_percent: default_jitter_percent(),
        }
    }
}

impl RetryPolicy {
    fn validate(&self, step_id: &str) -> Result<(), SpecError> {
        if self.multiplier == 0 || self.jitter_percent > 100 {
            return Err(SpecError::InvalidRetry(step_id.to_owned()));
        }
        if self.max_backoff_ms < self.initial_backoff_ms {
            return Err(SpecError::InvalidRetry(step_id.to_owned()));
        }
        Ok(())
    }
}

fn default_initial_backoff_ms() -> u64 {
    100
}

fn default_max_backoff_ms() -> u64 {
    60_000
}

fn default_multiplier() -> u32 {
    2
}

fn default_jitter_percent() -> u8 {
    20
}

#[derive(Debug, Error, PartialEq, Eq)]
pub enum SpecError {
    #[error("unsupported run spec version {0} (this kernel reads {SPEC_VERSION})")]
    UnsupportedVersion(String),
    #[error("unknown field \"{field}\" at {at} — refusing to guess (fail closed)")]
    UnknownField { at: String, field: String },
    #[error("malformed run spec: {0}")]
    Malformed(String),
    #[error("flow cli cannot be empty")]
    EmptyCli,
    #[error("trigger {0} must declare a non-empty id and executor")]
    InvalidTrigger(String),
    #[error("duplicate trigger id: {0}")]
    DuplicateTrigger(String),
    #[error("step id cannot be empty")]
    EmptyStepId,
    #[error("step {0} cli cannot be empty")]
    EmptyStepCli(String),
    #[error("duplicate step id: {0}")]
    DuplicateStep(String),
    #[error("step {0} must allow at least one iteration")]
    ZeroIterations(String),
    #[error("step {step} depends on unknown step {dependency}")]
    UnknownDependency { step: String, dependency: String },
    #[error("dependency cycle includes step {0}")]
    DependencyCycle(String),
    #[error("invalid retry policy for step {0}")]
    InvalidRetry(String),
}

#[cfg(test)]
mod tests;
